host-apis/wasi-0.2.0/host_api.cpp

```diff
@@ -574,31 +574,69 @@ void HttpOutgoingBody::write(const uint8_t *bytes, size_t len) {
   MOZ_RELEASE_ASSERT(write_to_outgoing_body(borrow, bytes, len));
 }
 
-Result<Void> HttpOutgoingBody::write_all(const uint8_t *bytes, size_t len) {
-  if (!valid()) {
-    // TODO: proper error handling for all 154 error codes.
-    return Result<Void>::err(154);
-  }
+class BodyWriteAllTask final : public api::AsyncTask {
+  HttpOutgoingBody *outgoing_body_;
+  PollableHandle outgoing_pollable_;
 
-  auto *state = static_cast<OutgoingBodyHandle *>(handle_state_.get());
-  Borrow<OutputStream> borrow(state->stream_handle_);
+  api::TaskCompletionCallback cb_;
+  Heap<JSObject *> cb_receiver_;
+  HostBytes bytes_;
+  size_t offset_ = 0;
 
-  while (len > 0) {
-    auto capacity_res = capacity();
-    if (capacity_res.is_err()) {
-      // TODO: proper error handling for all 154 error codes.
-      return Result<Void>::err(154);
+public:
+  explicit BodyWriteAllTask(HttpOutgoingBody *outgoing_body, HostBytes bytes,
+                          api::TaskCompletionCallback completion_callback,
+                          HandleObject callback_receiver)
+      : outgoing_body_(outgoing_body), cb_(completion_callback),
+        cb_receiver_(callback_receiver), bytes_(std::move(bytes)) {
+    outgoing_pollable_ = outgoing_body_->subscribe().unwrap();
+  }
+
+  [[nodiscard]] bool run(api::Engine *engine) override {
+    auto res = outgoing_body_->capacity();
+    if (res.is_err()) {
+      return false;
     }
-    auto capacity = capacity_res.unwrap();
-    auto bytes_to_write = std::min(len, static_cast<size_t>(capacity));
-    if (!write_to_outgoing_body(borrow, bytes, len)) {
-      return Result<Void>::err(154);
+    uint64_t capacity = res.unwrap();
+    MOZ_ASSERT(capacity >= 0);
+    auto bytes_to_write = std::min(bytes_.len - offset_, static_cast<size_t>(capacity));
+    outgoing_body_->write(bytes_.ptr.get() + offset_, bytes_to_write);
+    offset_ += bytes_to_write;
+    if (offset_ < bytes_.len) {
+      engine->queue_async_task(this);
```

Comment:
Is it possible for the capacity to increase after a write operation such that we should check the capacity again here and possibly loop?

My main concern would be in hosts where a low capacity is reported, that we might end up with a lot of async slowdown when sync loops should really be preferred?

Perhaps we should instead make the signal for polling an explicit capacity zero return per the WASI HTTP descriptions here?

Reply:
That's a great point, yes. I changed this to run in a loop until either all bytes have been written or no more capacity is signaled.